File: ts/src/session/action-labels.ts
/**
 * Compact action labels (AC-513 TS parity).
 */

import type { Coordinator, CoordinatorEvent } from "./coordinator.js";
import type { SessionEvent } from "./types.js";

const MAX_LABEL_LEN = 120;

const FAILURE_TYPES = new Set(["worker_failed", "turn_failed", "turn_interrupted", "session_failed", "session_canceled"]);

const EVENT_LABEL_MAP: Record<string, string> = {
  coordinator_created: "Coordinator started",
  worker_delegated: "Worker delegated",
  worker_completed: "Worker completed",
  worker_failed: "Worker failed",
  worker_redirected: "Worker redirected",
  fan_out: "Fan-out dispatched",
  fan_in: "Fan-in collected",
  session_created: "Session started",
  turn_submitted: "Turn submitted",
  turn_completed: "Turn completed",
  turn_interrupted: "Turn interrupted",
  turn_failed: "Turn failed",
};

function truncate(text: string): string {
  const clean = text.trim().replace(/\n/g, " ");
  if (clean.length <= MAX_LABEL_LEN) return clean;
  return clean.slice(0, MAX_LABEL_LEN - 1) + "…";
}

export class ActionLabel {
  readonly text: string;
  readonly category: string;

  constructor(text: string, category: string = "action") {
    this.text = text;
    this.category = category;
  }

  static create(text: string, category: string = "action"): ActionLabel {
    return new ActionLabel(truncate(text), category);
  }

  static noop(reason: string = "No changes"): ActionLabel {
    return new ActionLabel(truncate(reason), "noop");
  }
}

export function labelFromEvent(event: CoordinatorEvent | SessionEvent): ActionLabel {
  const eventType = event.eventType;
  const base = EVENT_LABEL_MAP[eventType] ?? eventType.replace(/_/g, " ");
  const payload = event.payload;
  const details: string[] = [];
  for (const key of ["task", "role", "reason", "error", "workerId", "turnId"]) {
    const val = payload[key];
    if (val) details.push(`${key}=${String(val).slice(0, 40)}`);
  }
  const text = details.length ? `${base}: ${details.slice(0, 3).join(", ")}` : base;
  const category = FAILURE_TYPES.has(eventType) ? "failure" : "action";
  return ActionLabel.create(text, category);
}

export function labelsFromCoordinator(coord: Coordinator, maxLabels: number = 20): ActionLabel[] {
  return coord.events.slice(-maxLabels).map(labelFromEvent);
}

File: autocontext/src/autocontext/session/action_labels.py
"""Compact action labels for timelines and event feeds (AC-513).

Domain concept: ActionLabel is a value object — a short, scannable
description of what just happened. Derived from events, not stored
as primary data.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from pydantic import BaseModel

from autocontext.session.coordinator import CoordinatorEventType
from autocontext.session.types import SessionEventType

if TYPE_CHECKING:
    from autocontext.session.coordinator import Coordinator, CoordinatorEvent
    from autocontext.session.types import SessionEvent

_MAX_LABEL_LEN = 120

_FAILURE_EVENT_TYPES = frozenset({
    CoordinatorEventType.WORKER_FAILED.value,
    SessionEventType.TURN_FAILED.value,
    SessionEventType.TURN_INTERRUPTED.value,
    SessionEventType.SESSION_FAILED.value,
    SessionEventType.SESSION_CANCELED.value,
})

_EVENT_LABEL_MAP: dict[str, str] = {
    "coordinator_created": "Coordinator started",
    "worker_delegated": "Worker delegated",
    "worker_started": "Worker started",
    "worker_completed": "Worker completed",
    "worker_failed": "Worker failed",
    "worker_redirected": "Worker redirected",
    "fan_out": "Fan-out dispatched",
    "fan_in": "Fan-in collected",
    "session_created": "Session started",
    "session_paused": "Session paused",
    "session_resumed": "Session resumed",
    "session_completed": "Session completed",
    "session_failed": "Session failed",
    "session_canceled": "Session canceled",
    "turn_submitted": "Turn submitted",
    "turn_completed": "Turn completed",
    "turn_interrupted": "Turn interrupted",
    "turn_failed": "Turn failed",
}


class ActionLabel(BaseModel):
    """Short, scannable description for timeline/event display.

    Categories: action, tool, failure, noop, info
    """

    text: str
    category: str = "action"

    @classmethod
    def create(cls, text: str, category: str = "action") -> ActionLabel:
        truncated = _truncate(text)
        return cls(text=truncated, category=category)

    @classmethod
    def noop(cls, reason: str = "No changes") -> ActionLabel:
        return cls(text=_truncate(reason), category="noop")

    model_config = {"frozen": True}


def _truncate(text: str) -> str:
    """Truncate to _MAX_LABEL_LEN with ellipsis."""
    text = text.strip().replace("\n", " ")
    if len(text) <= _MAX_LABEL_LEN:
        return text
    return text[: _MAX_LABEL_LEN - 1] + "…"


def label_from_event(event: CoordinatorEvent | SessionEvent) -> ActionLabel:
    """Derive a compact label from a coordinator or session event."""
    event_type = event.event_type.value
    base = _EVENT_LABEL_MAP.get(event_type, event_type.replace("_", " ").title())

    # Enrich with payload details
    payload = event.payload
    detail_parts: list[str] = []
    for key in ("task", "role", "reason", "error", "worker_id", "turn_id"):
        val = payload.get(key)
        if val:
            detail_parts.append(f"{key}={str(val)[:40]}")

    if detail_parts:
        # Keep labels glanceable in narrow timeline views rather than dumping every payload field.
        detail = ", ".join(detail_parts[:3])
        text = f"{base}: {detail}"
    else:
        text = base

    category = "failure" if event_type in _FAILURE_EVENT_TYPES else "action"
    return ActionLabel.create(text, category=category)


def labels_from_coordinator(
    coordinator: Coordinator,
    max_labels: int = 20,
) -> list[ActionLabel]:
    """Generate labels from the coordinator's recent events."""
    events = coordinator.events[-max_labels:]
    return [label_from_event(e) for e in events]
